mcs/class/System.Private.CoreLib/System/Type.cs

```diff
@@ -104,4 +117,386 @@ internal virtual string FormatTypeName (bool serialization)
+						else
+							type = type.MakeArrayType ();
+						break;
+					default:
+						type = type.MakeArrayType (mod);
+						break;
+					}
+				}
+			}
+
+			return type;
+		}
+
+		static Assembly ResolveAssembly (string name, Func<AssemblyName, Assembly> assemblyResolver, bool throwOnError,
+										 ref StackCrawlMark stackMark)
+		{
+			var aname = new AssemblyName (name);
+
+            if (assemblyResolver == null) {
+				if (throwOnError) {
+					return Assembly.Load (aname, ref stackMark, IntPtr.Zero);
+				} else {
+                    try {
+						return Assembly.Load (aname, ref stackMark, IntPtr.Zero);
+					} catch (FileNotFoundException) {
+						return null;
+					}
+				}
+			} else {
+				var assembly = assemblyResolver (aname);
+                if (assembly == null && throwOnError)
+                    throw new FileNotFoundException (SR.FileNotFound_ResolveAssembly, name);
+				return assembly;
+			}
+		}
+
+        static Type ResolveType (Assembly assembly, List<string> names, Func<Assembly, string, bool, Type> typeResolver, bool throwOnError, bool ignoreCase, ref StackCrawlMark stackMark)
+		{
+            Type type = null;
+
+            string name = EscapeTypeName (names [0]);
+            // Resolve the top level type.
+            if (typeResolver != null) {
+                type = typeResolver (assembly, name, ignoreCase);
+                if (type == null && throwOnError) {
+                    string error;
+					if (assembly == null)
+						throw new TypeLoadException (String.Format (SR.TypeLoad_ResolveType, name));
+					else
+						throw new TypeLoadException (String.Format (SR.TypeLoad_ResolveTypeFromAssembly, name, assembly.FullName));
+				}
+			} else {
+                if (assembly == null)
+                    type = RuntimeType.GetType (name, throwOnError, ignoreCase, false, ref stackMark);
+                else
+                    type = assembly.GetType (name, throwOnError, ignoreCase);
+			}
+
+			if (type == null)
+				return null;
+
+            // Resolve nested types.
+			BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Public;
+			if (ignoreCase)
+				bindingFlags |= BindingFlags.IgnoreCase;
+
+			for (int i = 1; i < names.Count; ++i) {
+				type = type.GetNestedType (names[i], bindingFlags);
+				if (type == null) {
+					if (throwOnError)
+						throw new TypeLoadException (String.Format (SR.TypeLoad_ResolveNestedType, names[i], names[i-1]));
+					else
+						break;
+                }
+			}
+			return type;
+		}
+
+        static string EscapeTypeName (string name)
+        {
+            if (name.IndexOfAny (SPECIAL_CHARS) < 0)
+                return name;
+
+			var sb = new StringBuilder ();
+            foreach (char c in name) {
+                if (Array.IndexOf<char> (SPECIAL_CHARS, c) >= 0)
+                    sb.Append ('\\');
+                sb.Append (c);
+            }
+
+			return sb.ToString ();
+        }
+
+		class ParsedName {
```

Comment:
Wouldn't be better to be a struct or ValueTuple?

Reply:
We use null for error checking.

Reply:
Use can still use nullable for that

Reply:
Thats doable, not sure its worth it, this is not really perf sensitive code.